hrms/hr/doctype/shift_assignment_tool/shift_assignment_tool.py

```diff
@@ -0,0 +1,245 @@
+		if self.to_date:
+			query = query.where(ShiftRequest.from_date <= self.to_date)
+
+		data = query.run(as_dict=True)
+		for d in data:
+			d.employee_name = d.employee + ": " + d.employee_name
+			d.shift_request = get_link_to_form("Shift Request", d.name)
+
+		return data
+
+	def get_query_for_employees_with_shifts(self):
+		ShiftAssignment = frappe.qb.DocType("Shift Assignment")
+		query = frappe.qb.from_(ShiftAssignment)
+
+		allow_multiple_shifts = frappe.db.get_single_value(
+			"HR Settings", "allow_multiple_shift_assignments"
+		)
+		# join Shift Type if multiple shifts are allowed as we need to know shift timings only in this case
+		if allow_multiple_shifts:
+			ShiftType = frappe.qb.DocType("Shift Type")
+			query = query.left_join(ShiftType).on(ShiftAssignment.shift_type == ShiftType.name)
+
+		query = (
+			query.select(ShiftAssignment.employee)
+			.distinct()
+			.where(
+				(ShiftAssignment.status == "Active")
+				& (ShiftAssignment.docstatus == 1)
+				# check for overlapping dates
+				& ((ShiftAssignment.end_date >= self.start_date) | (ShiftAssignment.end_date.isnull()))
+			)
+		)
+		if self.end_date:
+			query = query.where(ShiftAssignment.start_date <= self.end_date)
+
+		# check for overlapping timings if multiple shifts are allowed
+		if allow_multiple_shifts:
+			shift_start, shift_end = frappe.db.get_value(
+				"Shift Type", self.shift_type, ["start_time", "end_time"]
+			)
+			# turn it into a 48 hour clock for easier conditioning while considering overnight shifts
+			if shift_end < shift_start:
+				shift_end += timedelta(hours=24)
+			end_time_case = (
+				Case()
+				.when(ShiftType.end_time < ShiftType.start_time, ShiftType.end_time + Interval(hours=24))
+				.else_(ShiftType.end_time)
+			)
+			query = query.where((end_time_case >= shift_start) & (ShiftType.start_time <= shift_end))
+
+		return query
+
+	@frappe.whitelist()
+	def bulk_assign_shift(self, employees: list):
+		mandatory_fields = ["company", "shift_type", "start_date"]
+		validate_bulk_tool_fields(self, mandatory_fields, employees, "start_date", "end_date")
+
+		if len(employees) <= 30:
+			return self._bulk_assign_shift(employees)
+
+		frappe.enqueue(self._bulk_assign_shift, timeout=3000, employees=employees)
+		frappe.msgprint(
+			_("Creation of Shift Assignments has been queued. It may take a few minutes."),
+			alert=True,
+			indicator="blue",
+		)
+
+	def _bulk_assign_shift(self, employees: list):
+		success, failure = [], []
+		count = 0
+		savepoint = "before_shift_assignment"
+
+		for d in employees:
+			try:
+				frappe.db.savepoint(savepoint)
+				self.create_shift_assignment(d)
+			except Exception:
+				frappe.db.rollback(save_point=savepoint)
+				frappe.log_error(
+					f"Bulk Assignment - Shift Assignment failed for employee {d}.",
+					reference_doctype="Shift Assignment",
+				)
+				failure.append(d)
+			else:
+				success.append(d)
```

Comment:
should we show successful shift assignment links in the message too?

maybe a table format

<table>
<tr>
 <td>emp1
 <td>HR-SHIFT-1 (link)
<tr>
 <td>emp2
 <td>HR-SHIFT-2 (link)
</table>